apps/arweave/src/ar_repack_fsm.erl

```diff
@@ -109,24 +113,84 @@ next_state(#repack_chunk{state = needs_chunk} = RepackChunk) ->
-		{false, not_found, _, _} ->
+		{_, not_found, _, _} ->
 			%% This offset exists in some of the chunk indices, the chunk is not recorded
 			%% in the sync record. This can happen if there was some corruption at some
 			%% point in the past. We'll clean out the bad indices, and then record
 			%% the entropy.
 			invalid;
-		{false, TargetPacking, _, _} -> already_repacked;
-		{false, _, not_found, _} ->
+		{_, TargetPacking, _, _} -> already_repacked;
+		{_, _, not_found, _} ->
 			%% Chunk doesn't exist on disk, try chunk data db.
 			needs_data_path;
-		{false, _, _, false} -> 
+		{_, _, _, false} -> 
 			%% We are going to move this chunk to RocksDB after repacking so
 			%% we read its DataPath here to pass it later on to store_chunk.
 			needs_data_path;
-		{false, _, _, true} -> needs_repack;
-		_ ->
-			log_error(invalid_repack_fsm_transition, RepackChunk, []),
-			error
+		_ -> has_chunk
+	end,
+	RepackChunk#repack_chunk{state = NextState};
+
+%% ---------------------------------------------------------------------------
+%% State: needs_data_path
+%% ---------------------------------------------------------------------------
+next_state(#repack_chunk{
+		state = needs_data_path,
+		metadata = #chunk_metadata{data_path = not_set}} = RepackChunk) ->
+	%% Still waiting on data path.
+	RepackChunk;
+next_state(#repack_chunk{state = needs_data_path} = RepackChunk) ->
+	#repack_chunk{
+		chunk = Chunk,
+		metadata = Metadata
+	} = RepackChunk,
+	#chunk_metadata{
+		data_path = DataPath
+	} = Metadata,
+
+	IsInvalid = (
+		Chunk == not_found orelse
+		DataPath == not_found
+	),
+
+	NextState = case IsInvalid of
+		true -> 
+			%% This offset exists in some of the chunk indices and sync records, but there's
+			%% either no chunk data or no data_path. This can happen if there was some
+			%% corruption at some point in the past. We'll clean out the bad indices, and
+			%% then record the entropy.
+			invalid;
+		_ -> has_chunk
+	end,
+	RepackChunk#repack_chunk{state = NextState};
+
+%% ---------------------------------------------------------------------------
+%% State: has_chunk
+%% 
+%% has_chunk is an intermediate state to avoid duplicating state transition
+%% logic across both the needs_chunk and needs_data_path states. Once a chunk
+%% and optionally data_path have been read, we'll transition to has_chunk and
+%% then from there enter the repack logic.
+%% ---------------------------------------------------------------------------
+next_state(#repack_chunk{state = has_chunk} = RepackChunk) ->
+	#repack_chunk{
+		source_packing = SourcePacking,
+		target_packing = TargetPacking
+	} = RepackChunk,
+
+	NextState = case {SourcePacking, TargetPacking} of
+		_ when SourcePacking == TargetPacking ->
+			write_chunk;
+		{{replica_2_9, _}, _} ->
+			%% Source is replica_2_9, so we need its entropy first before we can unpacked it.
+			needs_source_entropy;
+		{unpacked_padded, {replica_2_9, _}} ->
+			needs_target_entropy;
```

Comment:
Why `needs_target_entropy` only when SourcePacking is unpacked_padded (and not unpacked or spora_2_6)?

Reply:
Ah, every source packing goes through unpacked_padded when the target packing is replica_2_9. The comments are very helpful, thank you.